File: ascii_fields/animations/plane.py
import math

from ..core import Animation, clamp, render_block_field, render_field, shade


WAVES = (
  (0.26, 1, 2, 1, 0.00),
  (0.21, 2, -1, -1, 0.23),
  (0.18, 3, 1, 2, 0.41),
  (0.15, -2, 3, -2, 0.67),
  (0.13, 4, -3, 3, 0.11),
  (0.10, 5, 2, -3, 0.52),
  (0.08, -4, -5, 4, 0.31),
  (0.06, 7, -2, -4, 0.79),
  (0.05, 6, 5, 5, 0.18),
)

RAMPS = {
  "clean": " .'`^\",:;Il!i~+_-?][}{1)(|\\/tfjrxnuvczXYUJCLQ0OZmwqpdbkhao*#MW&8%B@$",
  "soft": " .'`^\",:-_~+ito+xzMW#@$8",
  "dense": " .,:;irsXA253hMHGS#9B&@",
}


def _ramp_thresholds(ramp):
  n = len(ramp)
  return [(min(1.01, (i + 1) / n), ramp[i]) for i in range(n)]


class WavePlaneAnimation(Animation):
  """The original cyclic wave plane. Now char-rendered through the shared
  pipeline so themes apply and the full ramp can be used; pass --blocks to fall
  back to the smooth coloured-block look."""

  def render(self, width, height, elapsed, phase, options):
    grid = self._levels(width, height, phase, options)
    if options.blocks:
      return render_block_field(width, height, grid, options)
    ramp = RAMPS.get(options.charset, RAMPS["clean"])
    self.thresholds = _ramp_thresholds(ramp)
    return render_field(width, height, grid, options, self)

  def _levels(self, width, height, phase, options):
    grid = []
    for row in range(height):
      v = row / height
      line = []
      for col in range(width):
        u = col / width
        su, sv = self.sample_uv(u, v, phase, options.scroll)
        line.append(clamp(shade(self.wave_height(su, sv, phase, options.scale), options.contrast)))
      grid.append(line)
    return grid

  def sample_uv(self, u, v, phase, scroll):
    if not scroll:
      return u, v
    return (
      (u + 0.08 * math.cos(2.0 * math.pi * phase)) % 1.0,
      (v + 0.08 * math.sin(2.0 * math.pi * phase)) % 1.0,
    )

  def wave_height(self, u, v, phase, scale):
    x = 2.0 * math.pi * u
    y = 2.0 * math.pi * v
    t = 2.0 * math.pi * phase
    density = max(1, round(scale))
    z = 0.0
    for amplitude, wave_x, wave_y, time_speed, offset in WAVES:
      z += amplitude * math.sin(
        (wave_x * density) * x
        + (wave_y * density) * y
        + time_speed * t
        + offset * 2.0 * math.pi
      )
    return z

File: src/animations/plane.rs
//! Wave-plane: the original cyclic wave with a long scene-specific ASCII ramp.

use std::f64::consts::PI;
use crate::animation::{Animation, FrameContext};
use crate::core::{clamp, render_field, shade, FieldStyle};

const STYLE: FieldStyle = FieldStyle { gray_lo: 234, gray_hi: 255, default_theme: "ocean" };

// (amplitude, wave_x, wave_y, time_speed, offset)
const WAVES: &[(f64, f64, f64, f64, f64)] = &[
  (0.26, 1.0, 2.0, 1.0, 0.00),
  (0.21, 2.0, -1.0, -1.0, 0.23),
  (0.18, 3.0, 1.0, 2.0, 0.41),
  (0.15, -2.0, 3.0, -2.0, 0.67),
  (0.13, 4.0, -3.0, 3.0, 0.11),
  (0.10, 5.0, 2.0, -3.0, 0.52),
  (0.08, -4.0, -5.0, 4.0, 0.31),
  (0.06, 7.0, -2.0, -4.0, 0.79),
  (0.05, 6.0, 5.0, 5.0, 0.18),
];

const RAMP_CLEAN: &str = " .'`^\",:;Il!i~+_-?][}{1)(|\\/tfjrxnuvczXYUJCLQ0OZmwqpdbkhao*#MW&8%B@$";

fn ramp_thresholds(ramp: &str) -> Vec<(f64, char)> {
  let chars: Vec<char> = ramp.chars().collect();
  let n = chars.len();
  chars.iter().enumerate().map(|(i, &c)| {
    let limit = ((i + 1) as f64 / n as f64).min(1.01);
    (limit, c)
  }).collect()
}

pub struct WavePlane;

fn sample_uv(u: f64, v: f64, elapsed: f64, scroll: bool) -> (f64, f64) {
  if !scroll { return (u, v); }
  ((u + elapsed * 0.055).rem_euclid(1.0),
   (v + elapsed * 0.025).rem_euclid(1.0))
}

fn wave_height(u: f64, v: f64, phase: f64, scale: f64) -> f64 {
  let x = 2.0 * PI * u;
  let y = 2.0 * PI * v;
  let t = 2.0 * PI * phase;
  let density = scale.round().max(1.0);
  let mut z = 0.0_f64;
  for &(a, wx, wy, ts, off) in WAVES {
    z += a * ((wx * density) * x + (wy * density) * y + ts * t + off * 2.0 * PI).sin();
  }
  z
}

fn levels(ctx: &FrameContext) -> Vec<f64> {
  let w = ctx.width; let h = ctx.height;
  let mut g = vec![0.0_f64; w * h];
  let contrast = ctx.options.contrast;
  for row in 0..h {
    let v = row as f64 / h as f64;
    let base = row * w;
    for col in 0..w {
      let u = col as f64 / w as f64;
      let (su, sv) = sample_uv(u, v, ctx.elapsed, ctx.options.scroll);
      g[base + col] = clamp(shade(wave_height(su, sv, ctx.phase, ctx.options.scale), contrast));
    }
  }
  g
}

impl Animation for WavePlane {
  fn render(&mut self, ctx: &FrameContext, out: &mut String) {
    let g = levels(ctx);
    let th = ramp_thresholds(RAMP_CLEAN);
    render_field(ctx, &g, &th, &STYLE, out);
  }
}
